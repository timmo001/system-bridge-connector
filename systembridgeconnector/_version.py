"""
Provides systembridgeconnector version information.
"""

# This file is auto-generated! Do not edit!
# Use `python -m incremental.update systembridgeconnector` to change this file.

from incremental import Version

__version__ = Version("systembridgeconnector", 4, 0, 6, dev=0)
__all__ = ["__version__"]
