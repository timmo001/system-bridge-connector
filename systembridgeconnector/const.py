"""Constants."""

from enum import StrEnum


class QueryParameter(StrEnum):
    """Query Parameter."""

    ALBUM = "album"
    API_PORT = "apiPort"
    ARTIST = "artist"
    AUTOPLAY = "autoplay"
    BASE = "base"
    FILENAME = "filename"
    PATH = "path"
    TITLE = "title"
    TOKEN = "token"
    URL = "url"
    VOLUME = "volume"


class EventKey(StrEnum):
    """Event Key."""

    APP_ICON = "app_icon"
    APP_NAME = "app_name"
    BASE = "base"
    DATA = "data"
    DIRECTORIES = "directories"
    EVENT = "event"
    FILE = "file"
    FILENAME = "filename"
    FILES = "files"
    ID = "id"
    KEY = "key"
    MESSAGE = "message"
    MODULE = "module"
    MODULES = "modules"
    PATH = "path"
    SETTING = "setting"
    SUBTYPE = "subtype"
    TEXT = "text"
    TIMEOUT = "timeout"
    TITLE = "title"
    TOKEN = "token"
    TYPE = "type"
    URL = "url"
    VALUE = "value"
    VERSION = "version"
    VERSIONS = "versions"


class EventType(StrEnum):
    """Event Type."""

    APPLICATION_UPDATE = "APPLICATION_UPDATE"
    APPLICATION_UPDATING = "APPLICATION_UPDATING"
    DATA_GET = "DATA_GET"
    DATA_LISTENER_REGISTERED = "DATA_LISTENER_REGISTERED"
    DATA_LISTENER_UNREGISTERED = "DATA_LISTENER_UNREGISTERED"
    DATA_UPDATE = "DATA_UPDATE"
    DIRECTORIES = "DIRECTORIES"
    ERROR = "ERROR"
    EXIT_APPLICATION = "EXIT_APPLICATION"
    FILE = "FILE"
    FILES = "FILES"
    GET_DATA = "GET_DATA"
    GET_DIRECTORIES = "GET_DIRECTORIES"
    GET_FILE = "GET_FILE"
    GET_FILES = "GET_FILES"
    GET_SETTINGS = "GET_SETTINGS"
    KEYBOARD_KEY_PRESSED = "KEYBOARD_KEY_PRESSED"
    KEYBOARD_KEYPRESS = "KEYBOARD_KEYPRESS"
    KEYBOARD_TEXT = "KEYBOARD_TEXT"
    KEYBOARD_TEXT_SENT = "KEYBOARD_TEXT_SENT"
    MEDIA_CONTROL = "MEDIA_CONTROL"
    NOTIFICATION = "NOTIFICATION"
    NOTIFICATION_SENT = "NOTIFICATION_SENT"
    OPEN = "OPEN"
    OPENED = "OPENED"
    POWER_HIBERNATE = "POWER_HIBERNATE"
    POWER_HIBERNATING = "POWER_HIBERNATING"
    POWER_LOCK = "POWER_LOCK"
    POWER_LOCKING = "POWER_LOCKING"
    POWER_LOGGINGOUT = "POWER_LOGGINGOUT"
    POWER_LOGOUT = "POWER_LOGOUT"
    POWER_RESTART = "POWER_RESTART"
    POWER_RESTARTING = "POWER_RESTARTING"
    POWER_SHUTDOWN = "POWER_SHUTDOWN"
    POWER_SHUTTINGDOWN = "POWER_SHUTTINGDOWN"
    POWER_SLEEP = "POWER_SLEEP"
    POWER_SLEEPING = "POWER_SLEEPING"
    REGISTER_DATA_LISTENER = "REGISTER_DATA_LISTENER"
    SETTINGS_UPDATED = "SETTINGS_UPDATED"
    SETTINGS_RESULT = "SETTINGS_RESULT"
    UNREGISTER_DATA_LISTENER = "UNREGISTER_DATA_LISTENER"
    UPDATE_SETTINGS = "UPDATE_SETTINGS"


class EventSubType(StrEnum):
    """Event SubType."""

    BAD_TOKEN = "BAD_TOKEN"
    BAD_DIRECTORY = "BAD_DIRECTORY"
    BAD_FILE = "BAD_FILE"
    BAD_JSON = "BAD_JSON"
    BAD_KEY = "MISSING_KEY"
    BAD_PATH = "BAD_PATH"
    BAD_REQUEST = "BAD_REQUEST"
    INVALID_ACTION = "INVALID_ACTION"
    LISTENER_ALREADY_REGISTERED = "LISTENER_ALREADY_REGISTERED"
    LISTENER_NOT_REGISTERED = "LISTENER_NOT_REGISTERED"
    MISSING_ACTION = "MISSING_ACTION"
    MISSING_TOKEN = "MISSING_TOKEN"
    MISSING_BASE = "MISSING_BASE"
    MISSING_KEY = "MISSING_KEY"
    MISSING_MODULES = "MISSING_MODULES"
    MISSING_PATH = "MISSING_PATH"
    MISSING_PATH_URL = "MISSING_PATH_URL"
    MISSING_SETTING = "MISSING_SETTING"
    MISSING_TEXT = "MISSING_TEXT"
    MISSING_TITLE = "MISSING_TITLE"
    MISSING_VALUE = "MISSING_VALUE"
    UNKNOWN_EVENT = "UNKNOWN_EVENT"
