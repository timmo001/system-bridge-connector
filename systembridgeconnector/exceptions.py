"""Exceptions."""


class AuthenticationException(BaseException):
    """Raise this when there is an authentication issue."""


class BadMessageException(BaseException):
    """Raise this when a bad message is sent."""


class BadRequestException(BaseException):
    """Raise this when a bad request is sent."""


class ConnectionClosedException(BaseException):
    """Raise this when connection is closed."""


class ConnectionErrorException(BaseException):
    """Raise this when error connecting."""
