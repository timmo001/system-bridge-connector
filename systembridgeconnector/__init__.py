"""System Bridge Connector."""

from ._version import __version__  # re-export for convenience

__all__ = ["__version__"]
