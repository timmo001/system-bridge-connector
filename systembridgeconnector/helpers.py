"""Helpers."""

from collections.abc import Callable
import datetime as dt
from datetime import UTC, datetime, timedelta
from functools import partial

from systembridgemodels.const import ModulesData
from systembridgemodels.modules.cpu import PerCPU
from systembridgemodels.modules.displays import Display
from systembridgemodels.modules.gpus import GPU

utcnow = partial(dt.datetime.now, UTC)


def battery_time_remaining(data: ModulesData) -> datetime | None:
    """Return the battery time remaining."""
    if data.battery and (battery_time := data.battery.time_remaining) is not None:
        return utcnow() + timedelta(seconds=battery_time)
    return None


def camera_in_use(data: ModulesData) -> bool | None:
    """Return if any camera is in use."""
    if data.system and data.system.camera_usage is not None:
        return len(data.system.camera_usage) > 0
    return None


def cpu_speed(data: ModulesData) -> float | None:
    """Return the CPU speed."""
    if (
        data.cpu
        and (cpu_frequency := data.cpu.frequency) is not None
        and (cpu_frequency.current) is not None
    ):
        return round(cpu_frequency.current / 1000, 2)
    return None


def with_per_cpu(func) -> Callable:
    """Wrap a function to ensure per CPU data is available."""

    def wrapper(data: ModulesData, index: int) -> float | None:
        """Wrap a function to ensure per CPU data is available."""
        if data.cpu and data.cpu.per_cpu is not None and index < len(data.cpu.per_cpu):
            return func(data.cpu.per_cpu[index])
        return None

    return wrapper


@with_per_cpu
def cpu_power_per_cpu(per_cpu: PerCPU) -> float | None:
    """Return CPU power per CPU."""
    return per_cpu.power


@with_per_cpu
def cpu_usage_per_cpu(per_cpu: PerCPU) -> float | None:
    """Return CPU usage per CPU."""
    return per_cpu.usage


def with_display(func) -> Callable:
    """Wrap a function to ensure a Display is available."""

    def wrapper(data: ModulesData, index: int) -> Display | None:
        """Wrap a function to ensure a Display is available."""
        if data.displays and index < len(data.displays):
            return func(data.displays[index])
        return None

    return wrapper


@with_display
def display_resolution_horizontal(display: Display) -> int | None:
    """Return the Display resolution horizontal."""
    return display.resolution_horizontal


@with_display
def display_resolution_vertical(display: Display) -> int | None:
    """Return the Display resolution vertical."""
    return display.resolution_vertical


@with_display
def display_refresh_rate(display: Display) -> float | None:
    """Return the Display refresh rate."""
    return display.refresh_rate


def with_gpu(func) -> Callable:
    """Wrap a function to ensure a GPU is available."""

    def wrapper(data: ModulesData, index: int) -> GPU | None:
        """Wrap a function to ensure a GPU is available."""
        if data.gpus and index < len(data.gpus):
            return func(data.gpus[index])
        return None

    return wrapper


@with_gpu
def gpu_core_clock_speed(gpu: GPU) -> float | None:
    """Return the GPU core clock speed."""
    return gpu.core_clock


@with_gpu
def gpu_fan_speed(gpu: GPU) -> float | None:
    """Return the GPU fan speed."""
    return gpu.fan_speed


@with_gpu
def gpu_memory_clock_speed(gpu: GPU) -> float | None:
    """Return the GPU memory clock speed."""
    return gpu.memory_clock


@with_gpu
def gpu_memory_free(gpu: GPU) -> float | None:
    """Return the free GPU memory."""
    return gpu.memory_free


@with_gpu
def gpu_memory_used(gpu: GPU) -> float | None:
    """Return the used GPU memory."""
    return gpu.memory_used


@with_gpu
def gpu_memory_used_percentage(gpu: GPU) -> float | None:
    """Return the used GPU memory percentage."""
    if (gpu.memory_used) is not None and (gpu.memory_total) is not None:
        return round(gpu.memory_used / gpu.memory_total * 100, 2)
    return None


@with_gpu
def gpu_power_usage(gpu: GPU) -> float | None:
    """Return the GPU power usage."""
    return gpu.power_usage


@with_gpu
def gpu_temperature(gpu: GPU) -> float | None:
    """Return the GPU temperature."""
    return gpu.temperature


@with_gpu
def gpu_usage_percentage(gpu: GPU) -> float | None:
    """Return the GPU usage percentage."""
    return gpu.core_load


def memory_free(data: ModulesData) -> float | None:
    """Return the free memory."""
    if (
        data.memory
        and (virtual := data.memory.virtual) is not None
        and (free := virtual.free) is not None
    ):
        return round(free / 1000**3, 2)
    return None


def memory_used(data: ModulesData) -> float | None:
    """Return the used memory."""
    if (
        data.memory
        and (virtual := data.memory.virtual) is not None
        and (used := virtual.used) is not None
    ):
        return round(used / 1000**3, 2)
    return None


def partition_usage(
    data: ModulesData,
    device_index: int,
    partition_index: int,
) -> float | None:
    """Return the used memory."""
    if (
        data.disks
        and (devices := data.disks.devices) is not None
        and device_index < len(devices)
        and (partitions := devices[device_index].partitions) is not None
        and partition_index < len(partitions)
        and (usage := partitions[partition_index].usage) is not None
    ):
        return usage.percent
    return None
