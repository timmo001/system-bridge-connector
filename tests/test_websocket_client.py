"""Test the websocket client module."""

import asyncio
from typing import Any
from unittest.mock import patch

import aiohttp
import pytest

from systembridgeconnector.const import (
    EVENT_MODULES,
    TYPE_DATA_GET,
    TYPE_DATA_LISTENER_REGISTERED,
    TYPE_DIRECTORIES,
    TYPE_ERROR,
    TYPE_FILE,
    TYPE_FILES,
    TYPE_KEYBOARD_KEY_PRESSED,
    TYPE_KEYBOARD_TEXT_SENT,
    TYPE_NOTIFICATION_SENT,
    TYPE_OPENED,
    TYPE_POWER_HIBERNATING,
    TYPE_POWER_LOCKING,
    TYPE_POWER_LOGGINGOUT,
    TYPE_POWER_RESTARTING,
    TYPE_POWER_SHUTTINGDOWN,
    TYPE_POWER_SLEEPING,
)
from systembridgeconnector.exceptions import (
    ConnectionClosedException,
    ConnectionErrorException,
)
from systembridgeconnector.websocket_client import WebSocketClient
from systembridgemodels.const import MODEL_SYSTEM
from systembridgemodels.keyboard_key import KeyboardKey
from systembridgemodels.keyboard_text import KeyboardText
from systembridgemodels.media_control import MediaControl
from systembridgemodels.media_directories import MediaDirectory
from systembridgemodels.media_files import MediaFile, MediaFiles
from systembridgemodels.media_get_file import MediaGetFile
from systembridgemodels.media_get_files import MediaGetFiles
from systembridgemodels.modules import GetData, ModulesData, RegisterDataListener
from systembridgemodels.notification import Notification
from systembridgemodels.open_path import OpenPath
from systembridgemodels.open_url import OpenUrl
from systembridgemodels.response import Response
from systembridgemodels.update import Update

from . import API_HOST, API_PORT, REQUEST_ID, TOKEN, WebSocketGenerator

base_response = Response(
    id=REQUEST_ID,
    type="TEST",
    data={"test": "test"},
)

modules_data = ModulesData()


async def _get_websocket_client(
    ws_client: WebSocketGenerator,
    response: Response,
) -> WebSocketClient:
    """Return a websocket client."""
    ws = await ws_client(response=response)

    websocket_client = WebSocketClient(
        API_HOST,
        API_PORT,
        TOKEN,
        session=ws.client.session,
        websocket=ws,
    )

    async def _handle_module_data(
        module: str,
        data: Any,
    ) -> None:
        """Handle module data."""
        setattr(modules_data, module, data)

    # Run the listener in the background
    asyncio.get_event_loop().create_task(
        websocket_client.listen(callback=_handle_module_data),
        name="WebSocket Listener",
    )

    return websocket_client


@pytest.mark.asyncio
async def test_connection_error(ws_client: WebSocketGenerator):
    """Test connection error."""
    websocket_client = await _get_websocket_client(ws_client, base_response)

    with patch(
        "aiohttp.ClientSession.ws_connect",
        side_effect=aiohttp.ClientConnectionError,
    ), pytest.raises(ConnectionErrorException):
        await websocket_client.connect()


@pytest.mark.asyncio
async def test_connection_closed(ws_client: WebSocketGenerator):
    """Test connection closed."""
    websocket_client = await _get_websocket_client(ws_client, base_response)
    assert websocket_client.connected is True

    await websocket_client.close()
    assert websocket_client.connected is False

    with patch(
        "aiohttp.ClientSession.ws_connect",
        side_effect=aiohttp.ClientConnectionError,
    ), pytest.raises(ConnectionClosedException):
        await websocket_client.application_update(Update(version="0.0.0"))


@pytest.mark.asyncio
async def test_timeout(ws_client: WebSocketGenerator):
    """Test timeout."""
    websocket_client = await _get_websocket_client(ws_client, base_response)
    assert websocket_client.connected is True

    with patch(
        "systembridgeconnector.websocket_client.WebSocketClient._wait_for_future",
        side_effect=asyncio.TimeoutError,
    ):
        response = await websocket_client.get_data(
            GetData(
                modules=[MODEL_SYSTEM],
            ),
            request_id=REQUEST_ID,
        )

    assert isinstance(response, Response)
    assert response.id == REQUEST_ID
    assert response.type == TYPE_ERROR


@pytest.mark.asyncio
async def test_close(ws_client: WebSocketGenerator):
    """Test close."""
    websocket_client = await _get_websocket_client(ws_client, base_response)
    assert websocket_client.connected is True

    await websocket_client.close()
    assert websocket_client.connected is False


@pytest.mark.asyncio
async def test_application_update(ws_client: WebSocketGenerator):
    """Test application update."""
    websocket_client = await _get_websocket_client(ws_client, base_response)
    assert websocket_client.connected is True
    response = await websocket_client.application_update(
        Update(
            version="0.0.0",
        )
    )
    assert isinstance(response, Response)
    assert response.type == "N/A"
    assert response.data == {}


@pytest.mark.asyncio
async def test_exit_backend(ws_client: WebSocketGenerator):
    """Test exit backend."""
    websocket_client = await _get_websocket_client(ws_client, base_response)
    assert websocket_client.connected is True
    response = await websocket_client.exit_backend(
        request_id=REQUEST_ID,
    )
    assert isinstance(response, Response)
    assert response.id == REQUEST_ID
    assert response.type == "N/A"
    assert response.data == {}


@pytest.mark.asyncio
async def test_get_data(ws_client: WebSocketGenerator):
    """Test get data."""
    websocket_client = await _get_websocket_client(
        ws_client,
        Response(
            id=REQUEST_ID,
            type=TYPE_DATA_GET,
            data={EVENT_MODULES: [MODEL_SYSTEM]},
        ),
    )
    assert websocket_client.connected is True
    response = await websocket_client.get_data(
        GetData(
            modules=[MODEL_SYSTEM],
        ),
        request_id=REQUEST_ID,
    )
    assert isinstance(response, Response)
    assert response.id == REQUEST_ID
    assert response.type == TYPE_DATA_GET
    assert response.data == {EVENT_MODULES: [MODEL_SYSTEM]}

    # # Test wait for data
    # while modules_data.system is None:
    #     await asyncio.sleep(0.1)

    # assert modules_data.system is not None


@pytest.mark.asyncio
async def test_get_directories(ws_client: WebSocketGenerator):
    """Test get directories."""
    websocket_client = await _get_websocket_client(
        ws_client,
        Response(
            id=REQUEST_ID,
            type=TYPE_DIRECTORIES,
            data=[{"key": "documents", "path": "/documents"}],
        ),
    )
    assert websocket_client.connected is True
    response = await websocket_client.get_directories(
        request_id=REQUEST_ID,
    )
    assert isinstance(response, list)
    assert len(response) == 1
    assert isinstance(response[0], MediaDirectory)
    assert response[0].key == "documents"
    assert response[0].path == "/documents"


@pytest.mark.asyncio
async def test_get_files(ws_client: WebSocketGenerator):
    """Test get files."""
    websocket_client = await _get_websocket_client(
        ws_client,
        Response(
            id=REQUEST_ID,
            type=TYPE_FILES,
            data={
                "files": [
                    {
                        "name": "test",
                        "path": "path/to",
                        "fullpath": "path/to/test",
                        "size": 0,
                        "last_accessed": 0,
                        "created": 0,
                        "modified": 0,
                        "is_directory": False,
                        "is_file": True,
                        "is_link": False,
                        "mime_type": None,
                    }
                ],
                "path": "path/to",
            },
        ),
    )
    assert websocket_client.connected is True
    response = await websocket_client.get_files(
        MediaGetFiles(
            base="documents",
            path="path/to",
        ),
        request_id=REQUEST_ID,
    )
    assert isinstance(response, MediaFiles)
    assert response.files[0].name == "test"
    assert response.files[0].path == "path/to"
    assert response.files[0].fullpath == "path/to/test"
    assert response.files[0].size == 0
    assert response.files[0].last_accessed == 0
    assert response.files[0].created == 0
    assert response.files[0].modified == 0
    assert response.files[0].is_directory is False
    assert response.files[0].is_file is True
    assert response.files[0].is_link is False
    assert response.files[0].mime_type is None
    assert response.path == "path/to"


@pytest.mark.asyncio
async def test_get_file(ws_client: WebSocketGenerator):
    """Test get file."""
    websocket_client = await _get_websocket_client(
        ws_client,
        Response(
            id=REQUEST_ID,
            type=TYPE_FILE,
            data={
                "name": "test",
                "path": "path/to",
                "fullpath": "path/to/test",
                "size": 0,
                "last_accessed": 0,
                "created": 0,
                "modified": 0,
                "is_directory": False,
                "is_file": True,
                "is_link": False,
                "mime_type": None,
            },
        ),
    )
    assert websocket_client.connected is True
    response = await websocket_client.get_file(
        MediaGetFile(
            base="documents",
            path="path/to/test",
        ),
        request_id=REQUEST_ID,
    )
    assert isinstance(response, MediaFile)
    assert response.name == "test"
    assert response.path == "path/to"
    assert response.fullpath == "path/to/test"
    assert response.size == 0
    assert response.last_accessed == 0
    assert response.created == 0
    assert response.modified == 0
    assert response.is_directory is False
    assert response.is_file is True
    assert response.is_link is False
    assert response.mime_type is None


@pytest.mark.asyncio
async def test_register_data_listener(ws_client: WebSocketGenerator):
    """Test register data listener."""
    websocket_client = await _get_websocket_client(
        ws_client,
        Response(
            id=REQUEST_ID,
            type=TYPE_DATA_LISTENER_REGISTERED,
            data={EVENT_MODULES: [MODEL_SYSTEM]},
        ),
    )
    assert websocket_client.connected is True
    response = await websocket_client.register_data_listener(
        RegisterDataListener(
            modules=[MODEL_SYSTEM],
        ),
        request_id=REQUEST_ID,
    )
    assert isinstance(response, Response)
    assert response.id == REQUEST_ID
    assert response.type == TYPE_DATA_LISTENER_REGISTERED
    assert response.data == {EVENT_MODULES: [MODEL_SYSTEM]}


@pytest.mark.asyncio
async def test_keyboard_keypress(ws_client: WebSocketGenerator):
    """Test keyboard keypress."""
    websocket_client = await _get_websocket_client(
        ws_client,
        Response(
            id=REQUEST_ID,
            type=TYPE_KEYBOARD_KEY_PRESSED,
            data={"key": "a"},
        ),
    )
    assert websocket_client.connected is True
    response = await websocket_client.keyboard_keypress(
        KeyboardKey(
            key="a",
        ),
        request_id=REQUEST_ID,
    )
    assert isinstance(response, Response)
    assert response.id == REQUEST_ID
    assert response.type == TYPE_KEYBOARD_KEY_PRESSED
    assert response.data == {"key": "a"}


@pytest.mark.asyncio
async def test_keyboard_text(ws_client: WebSocketGenerator):
    """Test keyboard text."""
    websocket_client = await _get_websocket_client(
        ws_client,
        Response(
            id=REQUEST_ID,
            type=TYPE_KEYBOARD_TEXT_SENT,
            data={"text": "test"},
        ),
    )
    assert websocket_client.connected is True
    response = await websocket_client.keyboard_text(
        KeyboardText(
            text="test",
        ),
        request_id=REQUEST_ID,
    )
    assert isinstance(response, Response)
    assert response.id == REQUEST_ID
    assert response.type == TYPE_KEYBOARD_TEXT_SENT
    assert response.data == {"text": "test"}


@pytest.mark.asyncio
async def test_media_control(ws_client: WebSocketGenerator):
    """Test media control."""
    websocket_client = await _get_websocket_client(
        ws_client,
        Response(
            id=REQUEST_ID,
            type="N/A",
            data={},
        ),
    )
    assert websocket_client.connected is True
    response = await websocket_client.media_control(
        MediaControl(
            action="play",
        ),
        request_id=REQUEST_ID,
    )
    assert isinstance(response, Response)
    assert response.id == REQUEST_ID
    assert response.type == "N/A"
    assert response.data == {}


@pytest.mark.asyncio
async def test_send_notification(ws_client: WebSocketGenerator):
    """Test send notification."""
    websocket_client = await _get_websocket_client(
        ws_client,
        Response(
            id=REQUEST_ID,
            type=TYPE_NOTIFICATION_SENT,
            data={"title": "test", "message": "test"},
        ),
    )
    assert websocket_client.connected is True
    response = await websocket_client.send_notification(
        Notification(
            title="test",
            message="test",
        ),
        request_id=REQUEST_ID,
    )
    assert isinstance(response, Response)
    assert response.id == REQUEST_ID
    assert response.type == TYPE_NOTIFICATION_SENT
    assert response.data == {"title": "test", "message": "test"}


@pytest.mark.asyncio
async def test_open_path(ws_client: WebSocketGenerator):
    """Test open path."""
    websocket_client = await _get_websocket_client(
        ws_client,
        Response(
            id=REQUEST_ID,
            type=TYPE_OPENED,
            data={"path": "test"},
        ),
    )
    assert websocket_client.connected is True
    response = await websocket_client.open_path(
        OpenPath(
            path="test",
        ),
        request_id=REQUEST_ID,
    )
    assert isinstance(response, Response)
    assert response.id == REQUEST_ID
    assert response.type == TYPE_OPENED
    assert response.data == {"path": "test"}


@pytest.mark.asyncio
async def test_open_url(ws_client: WebSocketGenerator):
    """Test open url."""
    websocket_client = await _get_websocket_client(
        ws_client,
        Response(
            id=REQUEST_ID,
            type=TYPE_OPENED,
            data={"url": "test"},
        ),
    )
    assert websocket_client.connected is True
    response = await websocket_client.open_url(
        OpenUrl(
            url="test",
        ),
        request_id=REQUEST_ID,
    )
    assert isinstance(response, Response)
    assert response.id == REQUEST_ID
    assert response.type == TYPE_OPENED
    assert response.data == {"url": "test"}


@pytest.mark.asyncio
async def test_power_control(ws_client: WebSocketGenerator):
    """Test power control."""
    websocket_client = await _get_websocket_client(
        ws_client,
        Response(
            id=REQUEST_ID,
            type=TYPE_POWER_SLEEPING,
            data={},
        ),
    )
    assert websocket_client.connected is True
    response = await websocket_client.power_sleep(
        request_id=REQUEST_ID,
    )
    assert isinstance(response, Response)
    assert response.id == REQUEST_ID
    assert response.type == TYPE_POWER_SLEEPING
    assert response.data == {}


@pytest.mark.asyncio
async def test_power_hibernate(ws_client: WebSocketGenerator):
    """Test power control."""
    websocket_client = await _get_websocket_client(
        ws_client,
        Response(
            id=REQUEST_ID,
            type=TYPE_POWER_HIBERNATING,
            data={},
        ),
    )
    assert websocket_client.connected is True
    response = await websocket_client.power_hibernate(
        request_id=REQUEST_ID,
    )
    assert isinstance(response, Response)
    assert response.id == REQUEST_ID
    assert response.type == TYPE_POWER_HIBERNATING
    assert response.data == {}


@pytest.mark.asyncio
async def test_power_restart(ws_client: WebSocketGenerator):
    """Test power control."""
    websocket_client = await _get_websocket_client(
        ws_client,
        Response(
            id=REQUEST_ID,
            type=TYPE_POWER_RESTARTING,
            data={},
        ),
    )
    assert websocket_client.connected is True
    response = await websocket_client.power_restart(
        request_id=REQUEST_ID,
    )
    assert isinstance(response, Response)
    assert response.id == REQUEST_ID
    assert response.type == TYPE_POWER_RESTARTING
    assert response.data == {}


@pytest.mark.asyncio
async def test_power_shutdown(ws_client: WebSocketGenerator):
    """Test power control."""
    websocket_client = await _get_websocket_client(
        ws_client,
        Response(
            id=REQUEST_ID,
            type=TYPE_POWER_SHUTTINGDOWN,
            data={},
        ),
    )
    assert websocket_client.connected is True
    response = await websocket_client.power_shutdown(
        request_id=REQUEST_ID,
    )
    assert isinstance(response, Response)
    assert response.id == REQUEST_ID
    assert response.type == TYPE_POWER_SHUTTINGDOWN
    assert response.data == {}


@pytest.mark.asyncio
async def test_power_lock(ws_client: WebSocketGenerator):
    """Test power control."""
    websocket_client = await _get_websocket_client(
        ws_client,
        Response(
            id=REQUEST_ID,
            type=TYPE_POWER_LOCKING,
            data={},
        ),
    )
    assert websocket_client.connected is True
    response = await websocket_client.power_lock(
        request_id=REQUEST_ID,
    )
    assert isinstance(response, Response)
    assert response.id == REQUEST_ID
    assert response.type == TYPE_POWER_LOCKING
    assert response.data == {}


@pytest.mark.asyncio
async def test_power_logout(ws_client: WebSocketGenerator):
    """Test power control."""
    websocket_client = await _get_websocket_client(
        ws_client,
        Response(
            id=REQUEST_ID,
            type=TYPE_POWER_LOGGINGOUT,
            data={},
        ),
    )
    assert websocket_client.connected is True
    response = await websocket_client.power_logout(
        request_id=REQUEST_ID,
    )
    assert isinstance(response, Response)
    assert response.id == REQUEST_ID
    assert response.type == TYPE_POWER_LOGGINGOUT
    assert response.data == {}
