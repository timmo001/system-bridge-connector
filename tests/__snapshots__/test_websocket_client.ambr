# serializer version: 1
# name: test_application_update
  Response(id='test', type='N/A', data={}, subtype=None, message='Message sent', module=None)
# ---
# name: test_bad_token
  Response(id='test', type='N/A', data={}, subtype=None, message='Message sent', module=None)
# ---
# name: test_exit_backend
  Response(id='test', type='N/A', data={}, subtype=None, message='Message sent', module=None)
# ---
# name: test_get_data
  ModulesData(battery=None, cpu=None, disks=None, displays=None, gpus=None, media=None, memory=None, networks=None, processes=None, sensors=None, system=System(boot_time=1234, fqdn='hostname.local', hostname='hostname', ip_address_4='192.168.1.100', mac_address='00:00:00:00:00:00', platform_version='1.0.0', platform='platform', uptime=1234, users=[{'name': 'username', 'active': True, 'terminal': 'terminal', 'host': 'host', 'started': 1234, 'pid': 1234}], uuid='uuid', version='1.0.0', camera_usage=['camera1', 'camera2'], ip_address_6='::1', pending_reboot=True, version_latest='4.99.0', version_newer_available=True))
# ---
# name: test_get_directories
  list([
  ])
# ---
# name: test_get_file
  MediaFile(name='file1', path='path/to/file1', fullpath='/full/path/to/file1', size=100, last_accessed=100.0, created=100.0, modified=100.0, is_directory=False, is_file=True, is_link=False, mime_type='text/plain')
# ---
# name: test_get_files
  MediaFiles(files=[], path='/home/user/documents')
# ---
# name: test_keyboard_keypress
  Response(id='test', type='KEYBOARD_KEY_PRESSED', data={'key': 'a'}, subtype=None, message=None, module=None)
# ---
# name: test_keyboard_text
  Response(id='test', type='KEYBOARD_TEXT_SENT', data={'text': 'test'}, subtype=None, message=None, module=None)
# ---
# name: test_media_control
  Response(id='test', type='N/A', data={}, subtype=None, message='Message sent', module=None)
# ---
# name: test_open_path
  Response(id='test', type='OPENED', data={'path': '/home/user/documents'}, subtype=None, message=None, module=None)
# ---
# name: test_open_url
  Response(id='test', type='OPENED', data={'url': 'https://www.google.com'}, subtype=None, message=None, module=None)
# ---
# name: test_power_hibernate
  Response(id='test', type='POWER_HIBERNATING', data={}, subtype=None, message=None, module=None)
# ---
# name: test_power_lock
  Response(id='test', type='POWER_LOCKING', data={}, subtype=None, message=None, module=None)
# ---
# name: test_power_logout
  Response(id='test', type='POWER_LOGGINGOUT', data={}, subtype=None, message=None, module=None)
# ---
# name: test_power_restart
  Response(id='test', type='POWER_RESTARTING', data={}, subtype=None, message=None, module=None)
# ---
# name: test_power_shutdown
  Response(id='test', type='POWER_SHUTTINGDOWN', data={}, subtype=None, message=None, module=None)
# ---
# name: test_power_sleep
  Response(id='test', type='POWER_SLEEPING', data={}, subtype=None, message=None, module=None)
# ---
# name: test_register_data_listener
  Response(id='test', type='DATA_LISTENER_REGISTERED', data={'modules': ['system']}, subtype=None, message=None, module=None)
# ---
# name: test_send_notification
  Response(id='test', type='NOTIFICATION_SENT', data={'title': 'Test', 'message': 'test', 'icon': None, 'image': None, 'actions': None, 'timeout': None, 'audio': None}, subtype=None, message=None, module=None)
# ---
# name: test_unknown_message
  Response(id='test', type='N/A', data={}, subtype=None, message='Message sent', module=None)
# ---
# name: test_wait_for_response_timeout
  ModulesData(battery=None, cpu=None, disks=None, displays=None, gpus=None, media=None, memory=None, networks=None, processes=None, sensors=None, system=System(boot_time=1234, fqdn='hostname.local', hostname='hostname', ip_address_4='192.168.1.100', mac_address='00:00:00:00:00:00', platform_version='1.0.0', platform='platform', uptime=1234, users=[{'name': 'username', 'active': True, 'terminal': 'terminal', 'host': 'host', 'started': 1234, 'pid': 1234}], uuid='uuid', version='1.0.0', camera_usage=['camera1', 'camera2'], ip_address_6='::1', pending_reboot=True, version_latest='4.99.0', version_newer_available=True))
# ---
